src/Microsoft.VisualStudio.Threading/NonConcurrentSynchronizationContext.cs

```diff
@@ -0,0 +1,178 @@
+    /// Once that delegate returns the next delegate can be executed.
+    /// </remarks>
+    public sealed class NonConcurrentSynchronizationContext : SynchronizationContext
+    {
+        /// <summary>
+        /// The queue of work to execute, if this is the original instance.
+        /// </summary>
+        private readonly AsyncQueue<(SendOrPostCallback, object?)>? queue;
+
+        /// <summary>
+        /// A value indicating whether to set this instance as <see cref="SynchronizationContext.Current" /> when invoking delegates.
+        /// </summary>
+        private readonly bool sticky;
+
+        /// <summary>
+        /// The original instance (on which <see cref="CreateCopy"/> was called).
+        /// </summary>
+        private readonly NonConcurrentSynchronizationContext? copyOf;
+
+        /// <summary>
+        /// Set to the <see cref="Environment.CurrentManagedThreadId"/> when a delegate is currently executing.
+        /// </summary>
+        private int? activeManagedThreadId;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NonConcurrentSynchronizationContext"/> class.
+        /// </summary>
+        /// <param name="sticky">
+        /// A value indicating whether to set this instance as <see cref="SynchronizationContext.Current" /> when invoking delegates.
+        /// This has the effect that async methods that are invoked on this <see cref="SynchronizationContext" />
+        /// will execute their continuations on this <see cref="SynchronizationContext" /> as well unless they use <see cref="Task.ConfigureAwait(bool)" /> with <c>false</c> as the argument.
+        /// </param>
+        public NonConcurrentSynchronizationContext(bool sticky)
+        {
+            this.queue = new AsyncQueue<(SendOrPostCallback, object?)>();
+            this.sticky = sticky;
+
+            // Start the queue processor. It will handle all exceptions.
+            this.ProcessQueueAsync().Forget();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NonConcurrentSynchronizationContext"/> class
+        /// that is a copy of an existing instance.
+        /// </summary>
+        /// <param name="copyFrom">The instance to copy from.</param>
+        private NonConcurrentSynchronizationContext(NonConcurrentSynchronizationContext copyFrom)
+        {
+            // We do *not* kick off processing of the queue, since the original copy has done that and we don't want two logical threads processing the queue.
+            this.sticky = copyFrom.sticky;
+            this.copyOf = copyFrom;
+        }
+
+        /// <summary>
+        /// Occurs when posted work throws an unhandled exception.
+        /// </summary>
+        public event EventHandler<Exception>? UnhandledException;
+
+        /// <inheritdoc />
+        public override void Post(SendOrPostCallback d, object? state)
+        {
+            if (this.copyOf is object)
```

Comment:
I keep seeing this with nullables? Is this a replacement for null check?

Reply:
It's orthogonal to nullable ref annotations. But it's preferred over `!=null` because it won't compile if the tested value is a value-type. So for example `3 != null` compiles but `3 is object` will not. So it helps avoid bugs where you are testing a member against null, but that member is a *struct* so it will never be null.